doc/source/serve/deployment-graph.md

```diff
@@ -0,0 +1,565 @@
+    return int(input_data)
+
+@serve.deployment
+async def avg_preprocessor(input_data):
+    """Simple feature processing that returns average of input list as float."""
+    time.sleep(0.15) # Manual delay for blocking computation
+    return sum(input_data) / len(input_data)
+
+# DAG building
+with InputNode() as dag_input:
+    # Partial access of user input by index
+    preprocessed_1 = preprocessor.bind(dag_input[0])  
+    preprocessed_2 = avg_preprocessor.bind(dag_input[1])
+```
+
++++
+
+There are two new APIs used in the DAG building stage: `InputNode()` and `bind()`.
+
+### **```InputNode()```** : User input of the graph
+
+```InputNode``` is a special singleton in DAG building that's only relevant to it's runtime call behavior. Even though all decorated classes or functions can be reused in arbitrary way to facilitate DAG building where the root DAGNode forms the graph with its children, in each deployment graph there should be one and only one InputNode used.
+
+```InputNode``` value is fulfilled and replaced by user input at runtime, therefore it takes no argument when being constructed.
+
+```InputNode``` should only be created within a context manager to indicate there's one and only one `InputNode` for each graph, and will throw exception if directly created, or multiple is found.
+
+It's possible to access partial user input by index or key, if a node in the graph doesn't need the complete user input to run. Example: User submits `[Tensor_0, Tensor_1]` via http but a model node might only need to receive `Tensor_1` at index 1 rather than complete list to reduce intermediate data transfer.
+
+
+### **`bind(*args, **kwargs)`** : The graph building API
+
+Once called on supported Ray-decorated function or class (`@serve.deployment` is fully supported, `@ray.remote` will be soon), generates a `DeploymentNode` of type `DAGNode` that acts as the building block of graph building.
+
+
+In the example above, we can see we're using a context manager to build and bind user input:
+```python
+with InputNode() as dag_input:
+```
+
+Which can be used and accessed by index or key in downstream calls of .bind(), such as:
+```python
+preprocessed_1 = preprocessor.bind(dag_input[0])
+```
+
+This means we're creating a DAGNode IR called `preprocessed_1` in graph building by calling `.bind()` on a serve decorated function, where it executes the decorated deployment function `preprocessor` that takes the user input at index 0 at runtime.
+
+#### bind() on function
+
+```bind()``` on function produces a DAGNode that can be executed with user input.
+
+```{tip}
+Each deployment node used in graph is individually scalable and configurable by default. This means in real production workload where we can expect difference in compute resource and latency, we can fine tune the nodes to optimal `num_replicas` and `num_cpus` to avoid a single node being the bottleneck of your deployment graph's latency or throughput.
```

Comment:
we should have a performance tuning section in the future!

Reply:
this is the beginning of where things get interesting ;)